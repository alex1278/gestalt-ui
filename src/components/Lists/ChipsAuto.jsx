import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Autocomplete } from 'react-md';
import { Error } from 'components/Typography';
import { insertItem, removeItem } from 'util/helpers/lists';
import Chip from './Chip';

const List = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

class Chips extends Component {
  static propTypes = {
    meta: PropTypes.object.isRequired,
    input: PropTypes.object,
    label: PropTypes.string,
    data: PropTypes.array,
    helpText: PropTypes.string,
  };

  static defaultProps = {
    input: {},
    label: 'Entry',
    data: [],
    helpText: null,
  };

  state = { touched: false };

  addItem = (value) => {
    const isDupe = this.props.input.value && this.props.input.value.find(item => item === value);

    if (value && !isDupe) {
      this.props.input.onChange(insertItem(this.props.input.value, value));
    }
  }

  removeItem = (value) => {
    this.props.input.onChange(removeItem(this.props.input.value, value));
  }

  handleTouched = () => {
    this.setState({ touched: true });
  }

  render() {
    const { label, data, helpText, input, meta: { error }, ...rest } = this.props;

    return (
      <React.Fragment>
        <Autocomplete
          id={`${label}-chips`}
          placeholder={this.props.label}
          fullWidth
          onAutocomplete={this.addItem}
          lineDirection="center"
          onFocus={this.handleTouched}
          filter={Autocomplete.caseInsensitiveFilter}
          data={data}
          clearOnAutocomplete
          focusInputOnAutocomplete
          listInline
          helpText={helpText}
          {...rest}
        />
        <List maxHeight="184px">
          {Array.isArray(input.value) && input.value.map((item, i) => <Chip key={`${item}--${i}`} item={item} onRemove={this.removeItem} />)}
        </List>
        {(this.state.touched && !!error) && <Error>{error}</Error>}
      </React.Fragment>
    );
  }
}

export default Chips;
