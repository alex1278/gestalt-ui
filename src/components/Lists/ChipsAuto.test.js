import React from 'react';
import { render } from 'react-testing-library';
import ChipsAuto from './ChipsAuto';

describe('ChipsAuto', () => {
  it('mounts with basic props', () => {
    const { container } = render(<ChipsAuto meta={{ error: null, touched: false }} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
