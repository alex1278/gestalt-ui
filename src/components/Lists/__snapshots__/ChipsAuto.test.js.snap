// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipsAuto mounts with basic props 1`] = `
<div
  aria-expanded="false"
  aria-haspopup="true"
  class="md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container"
  id="Entry-chips-menu"
>
  <div
    class="md-text-field-container md-full-width md-text-field-container--input md-autocomplete"
  >
    <input
      aria-autocomplete="list"
      autocomplete="off"
      class="md-text-field md-text-field--margin md-full-width md-text"
      id="Entry-chips"
      placeholder="Entry"
      type="text"
      value=""
    />
    <hr
      class="md-divider md-divider--text-field md-divider--expand-from-center"
    />
  </div>
</div>
`;
